<a href="https://colab.research.google.com/github/Rauloigs/portfolio/blob/main/tarea_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import numpy as np
import math

### Exercise 1
A sum of 9,000 paid into a bank account for two months (61 days) to attract simple interest will produce 9, 020 at the end of the term. Find the interest rate r and the return on this investment.

V(t) = $P*(1+r/m)$

P = $9000$

$V(t) = 9020 = P(1+(61/365)*r)$

$r = (365/61)((9020/900)-1)$

In [ ]:
print(f"{round((365/61)*((9020/9000)-1)*100,2)}%")

1.33%


### Exercise 2
How much would you pay today to receive 1, 000 at a certain future date if you require a return of 2% ? Note: It is not necessary to know the period of time because you have the return (2%)

V(t) = 1,000

r = 2%

V(t) = $P*(1+r/m)$

P = $V(t) / (1+r/m) $


In [ ]:
round(1000/(1+0.02),2)

980.39

### Exercise 3
How long will it take for a sum of 800 attracting simple interest to become $830 if the rate is 9%? Compute the return on this investment.

$V(t)=P(1+(r/m))$

V(t) = 830

P = 800

r = 9%

m = 12

1... $V(t) = P+P(rt/12)$

2... $t = (V(t) - P)/(Pr)*12$



In [ ]:
# 2...
print(f"tiempo = {round((830-800)/(800*0.09),2)*12} meses")

tiempo = 5.04 meses


Rendimiento

$K(s,t) = (V(t)-V(s)/ V(s))$

$V(s)$ es valor inicial


In [ ]:
print(f"Rendimiento: {(830-800)/(800)*100}%")

Rendimiento: 3.75%


### Exercise 4
Find the principal to be deposited initially in an account attracting simple interest at a rate of 8% if 1, 000 is needed after three months (91 days).

$V(t)=P(1+(rt/m))$

$P = ?$

$r = 0.08$

$V(t) = 1,000$

$t/m = 91/365$

1... $1,000 = P*(1+(0.08*(91/365)))$


In [ ]:
value_ex4 = (1+(0.08*(91/365)))
value_ex4

1.019945205479452

2... $1,000 = P*(1.01994)$

3... $P = 1,000/1.01994$

In [ ]:
print(f"The principal is: {round(1000/value_ex4,2)}")

The principal is: 980.44


### Exercise 5
Find and compare the future value after two years of a deposit of 100 attracting interest at a rate of 10% compounded a) annually and b) semi-annually.

$V(t)=P(1+(r/m))^{tm}$

$P=100$

$r=0.10$

a) $t/m = (1/1) = 1$

b) $t/m = 1/2 = 0.5$


In [ ]:
# a)
round(100*(1+(0.1*1))**(2*1),2)

121.0

In [ ]:
# b)
round(100*(1+(0.1*1/2))**(2*2),2)

121.55

### Exercise 6
Which will deliver a higher future value after one year, a deposit of 1, 000 attracting interest at 15% compounded daily, or at 15.5% compounded semiannually?

$V(t)=P(1+(rt/m))^{tm}$


In [ ]:
p=1000
r_daily=0.15
m_daily=365
r_semianually=0.155
m_semianually=2
t=1

In [ ]:
daily=p*(1+r_daily*(t/m_daily))**(m_daily)
semi_anually = p*(1+r_semianually*(t/m_semianually))**(m_semianually)

In [ ]:
print(f"Future Value semiannually compunded is:${f'{semi_anually:,.2f}'}\n")
print(f"Future Value daily compunded is:${f'{daily:,.2f}'} there for it delivers higher value")

Future Value semiannually compunded is:$1,161.01

Future Value daily compunded is:$1,161.80 there for it delivers higher value


### Exercise 7
What initial investment subject to annual compounding at 12% is needed to produce $1,000 after two years?

$V(t)=P(1+(rt/m))^{tm}$

$P=[V(t)]/[(1+(r/m))^{tm}]$

$P = ?$

In [ ]:
# datos
v_t=1000
t=2
m=1
r=0.12

In [ ]:
p = (v_t)/(1+(r/m))**(t*m)
print(f"{f'${p:,.2f}'}")

$797.19


## Exercise 8
Find the present value of 100, 000 to be received after 100 years if the interest rate is assumed to be 5% throughout the whole period and a) daily or b) annual compounding applies.

$P=[V(t)]/[(1+(r/m))^{tm}]$



In [ ]:
t=100
v_t=100000
r=0.05
#m_daily already declared, but want to have visibility...
m_daily=365
m_annual=1

In [ ]:
# a) daily
p = (v_t)/(1+(r/m_daily))**(t*m_daily)
print(f"The daily compunded Present Value is: ${round(p,2)}")

The daily compunded Present Value is: $674.03


In [ ]:
# b) annual
p = (v_t)/(1+(r/m_annual))**(t*m_annual)
print(f"The annual compunded Present Value is: ${round(p,2)}")

The annual compunded Present Value is: $760.45


## Exercise 9
Find the return over one year under monthly compounding with r = 10%.

In [ ]:
t=1
m=12
p=1
r=0.1

Asumiendo que $1 es nuestro P hay que obtener V(t)

$V(t)=P(1+(r/m))^{tm}$

In [ ]:
v_t=p*(1+(r/m))**(t*m)
print(f"The final value V(t) is: ${round(v_t,3)}")

The final value V(t) is: $1.105


Rendimiento

$K(s,t) = (V(t)-V(s)/ V(s))$

Para nuestro caso $p=V(s)$

In [ ]:
v_s=p

k_st=(v_t-v_s)/(v_s)*100
print(f"The return is: {round(k_st,2)}%")

The return is: 10.47%


### *Example of Annuity from Class*

Consider a loan of $1, 000 to be paid back in 5 equal instalments due at yearly intervals. The instalments include both the interest payable each year calculated at 15% of the current outstanding balance and the repayment of a fraction of the loan. A loan of this type is called an amortised loan. The amount of each instalment can be computed as
 1, 000/PA(15%, 5)= 298.32. This is because the loan is equivalent to an annuity from the point of view of the lender.

 $VP=A*[(1-(1+r)^(-n))/r]$

In [ ]:
n=5
r=0.15
p=1000

In [ ]:
# Solving for the right side...to get VP = A * x, where x is:
x=((1-(1+r)**(-n))/r)
x

3.352155098011401

In [ ]:
# Solving for Annuity
annuity=p/x
round(annuity,2)

298.32

In [ ]:
# VF after 5 years
round(annuity*5,2)

1491.58

## Exercise 10
How much can you borrow if the interest rate is 18%, you can afford to pay $10, 000 at the end of each year, and you want to clear the loan in 10 years?

$PA(r, n) = (1 − (1 + r)^{-n})/ r$

or... $VP=A*[(1-(1+r)^{-n})/r]$


In [ ]:
annuity=10000
r=0.18
n=10
# VP=?

In [ ]:
vp=annuity*((1-(1+r)**(-n))/r)
print(f"You can borrow: ${f'{vp:,.2f}'}")

You can borrow: $44,940.86


### Exercise 11
Suppose that you deposit $1, 200 at the end of each year for 40 years, subject to annual compounding at a constant rate of 5%. Find the balance after 40 years.

remember the Future Value: $V(t)=P(1+(r/m))^{tm}$

now, from this $VP=A*[(1-(1+r)^{-n})/r]$

to this:  $VF = A*[(1+r)^n-1]/r$ ... it is basically the same

In [ ]:
annuity=1200
n=40
r=0.05

In [ ]:
vf=annuity*(((1+r)**n)-1)/r

print(f"The Balance after 40 years will be: ${f'{vf:,.2f}'}")

The Balance after 40 years will be: $144,959.73


### Exercise 12
How long will it take to earn $1$ in interest if $1, 000, 000 is deposited at 10% compounded continuously?

$V(t) = r*e^{tr}*P = rV (t)$

$V(t) = Pe^{rt}$

In [ ]:
r=0.1
p=1000000
v_t=1000000+1
# t=?

start with this: $V(t) = Pe^{rt}$

1... $V(t)/ P = e^{rt}$

2... $ln(V(t)/ P) = ln(e^{rt})$

3... $ln(V(t)/ P) = rt*ln(e)$ ; thus $ln(e) = 1$

4... $ln(V(t)/ P) = rt$

5... $t = ln(V(t)/ P)/r$



In [ ]:
t = np.log(v_t/p)/r
print(f"To earn $1 with a principal of $1,000,000 it wil take: {t}, which is approximately one breath")

To earn $1 with a principal of $1,000,000 it wil take: 9.999994999180667e-06, which is approximately one breath


### Exercise 13
Find the rate for continuous compounding equivalent to monthly compounding at 12%.

For continuous compounding will use: $V(t) = Pe^{rt}$

For monthly compounding will use: $V(t)=P(1+(r/m))^{tm}$

Will also use:

$P=100$

$t=1$



First we should get for Monthly

In [ ]:
m=12
t=1
r=0.12
p=100

In [ ]:
v_t_monthly = p*(1+(r/m))**(t*m)
print(f"The future value for monthly compounded is: ${f'{v_t_monthly:,.2f}'}")

The future value for monthly compounded is: $112.68


Now we need to now the "r" at continuos compounding under the same conditions generating that V(t), in this case *v_t_monthly*

start with this: $V(t) = Pe^{rt}$

1... $V(t)/ P = e^{rt}$

2... $ln(V(t)/ P) = ln(e^{rt})$

3... $ln(V(t)/ P) = rt*ln(e)$ ; thus $ln(e) = 1$

4... $ln(V(t)/ P) = rt$

5... $r = ln(V(t)/ P)/t$

In [ ]:
r_continuos = (np.log(v_t_monthly/p)/t)*100
print(f"the rate for continuous compounding equivalent to monthly compounding at 12% is: {f'{r_continuos:,.2f}'}%")

the rate for continuous compounding equivalent to monthly compounding at 12% is: 11.94%


Now justo to validate with:

$V(t) = Pe^{rt}$

In [ ]:
v_t_continuos = p*(math.e)**(r*t)
print(f"With the calculated rate of: {f'{r_continuos:,.2f}'}%, The Final Value is: ${f'{v_t_continuos:,.2f}'}")
print(f"The future value for monthly compounded is: ${f'{v_t_monthly:,.2f}'}")

With the calculated rate of: 11.94%, The Final Value is: $112.75
The future value for monthly compounded is: $112.68


### Exercise 14
Find the frequency of periodic compounding at 20% to be equivalent to annual compounding at 21%

$V(t)=P(1+(r/m))^{tm}$

We will assume $P=1000$

In [ ]:
p=1000
r_annual=0.21
r_period=0.20
m_annual=1
t_annual=1

$Annual = Periodic$

$P(1+(r/m))^{tm} = P(1+(r/m))^{tm}$

First lets get de V(t) annual to later get the equivalent *m*

In [ ]:
v_t_annual = p*(1+(r_annual/m_annual)**(t_annual*m_annual))
print(f"The first side of the equation, which represents v(t) annual is: ${f'{v_t_annual:,.2f}'}")

The first side of the equation, which represents v(t) annual is: $1,210.00


Now we should clear for m

$1,210 = 1,000(1+(0.20/m))^{1*m}$

To isolate this is pretty difficult so I just substitute 2,3,4... n on the formula.

In this case you can see **m=2** to get $V(t) = $1,210$

**So it should compound by semester (2 times a year at 21%) to be equivalent**

In [ ]:
round(1000*(1+(0.2/2))**2,2)

1210.0